src/TraceEvent/Computers/StartStopActivityComputer.cs

```diff
@@ -957,53 +962,117 @@ void FixAndProcessAdoNetEvents(TraceEvent data)
                             }
-                        }
-                        first = false;
+                            OnStart(data, extraInfo, null, null, null, taskName);
+                            break;
+                        case "Microsoft.AspNetCore.Hosting.BeginRequest":
+                            // Just change the name from "Activity1Start" to "AspNetCoreHosting".
+                            OnStart(data, null, null, null, null, "AspNetCoreHosting");
+                            break;
                     }
-                    extraInfo = Utilities.StringBuilderCache.GetStringAndRelease(sb);
+                    result = true;
                 }
-                OnStart(data, extraInfo, null, null, null, taskName);
-                return true;
             }
             catch (Exception ex)
             {
                 Trace.WriteLine("Threw exception while processing DiagnosticSource start events: " + ex.Message);
             }
-            return false;
+
+            return result;
+        }
+
+        void FixAndProcessAppInsightsEvents(TraceEvent data)
+        {
+            Debug.Assert(data.ProviderGuid == MicrosoftApplicationInsighsDataProvider);
+            Debug.Assert(data.Opcode == TraceEventOpcode.Start || data.Opcode == TraceEventOpcode.Stop);
+
+            if (data.TaskName.Equals("Request", StringComparison.OrdinalIgnoreCase))
+            {
+                if (data.Opcode == TraceEventOpcode.Start)
+                {
+                    // If we have a related activity ID, then this start event
+                    // is already being tracked by an outer activity (e.g. an
+                    // ASP.Net request). Do not process it.
+                    if (!IsTrivialActivityId(data.RelatedActivityID))
+                    {
+                        return;
+                    }
+                    else
+                    {
+                        var extraInfo = (string)data.PayloadByName("Name") ?? (string)data.PayloadByName("Id");
+                        OnStart(data, extraInfo);
+                    }
+                }
+                else
+                {
+                    Debug.Assert(data.Opcode == TraceEventOpcode.Stop);
+                    OnStop(data);
+                }
+            }
+            else if (data.TaskName.Equals("Operation", StringComparison.OrdinalIgnoreCase))
+            {
+                // TODO: [BinDu] AppInsights SDK has an issue that can create unexpected nested activity for Operation task. 
+                // Ignore the events for now until the issue is fixed in AppInsights SDK 2.4.
+                // https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/540 
+                return;
+            }
+            else
+            {
+                Debug.Assert(false, $"{data.TaskName} is not recognized");
+                return;
+            }
+        }
+
+        bool IsTrivialActivityId(Guid g)
```

Comment:
Go ahead and make this static and move the declaration of GUID_ONE here.   If the language supported it we would make the static a local to this method, and placing them next to each other is our best approximation of that.

Reply:
Fixed in i2.